epistasis_selection.py

```diff
@@ -0,0 +1,98 @@
+import itertools
+from typing import Optional, Iterable, Tuple, List
+
+import numpy as np
+
+import potts_model
+import utils
+
+
+def combine_k_rounds(num_rounds: int, mutations: Iterable[Iterable[Tuple[int, int]]]) -> List[List[Tuple[int, int]]]:
+  """Return the result of combining `mutations` for `num_rounds`.
+
+  Starting with a pool of M `mutations` m_1 ... m_M, stack them for K=`num_rounds` rounds. For example,
+  for K=3 rounds of combination, this will result in every variant (m_i + m_j + m_k), for i, j, k \in M.
+  Be careful of memory usage, as this can be very large due to combinatorial possibilities.
+  In the best case, this scales with {M \choose K}. But if mutations overlap at P positions,
+  combining them produces 2^{P} variants. So in the worst case, this will produce
+  {M \choose K} * 2^{P} variants. See the definition for `utils.merge_mutation_sets` for more on
+  mutation merging.
+
+  Args:
+    num_rounds: The number of rounds of combination
+    mutations: The starting pool of mutations, where each mutation is an iterable of
+      tuples encoding mutations (position, mutation).
+
+  Returns:
+    A list of tuples of mutations, where each element will be a combination of
+    `num_rounds` mutations from `mutations`. Note that each tuple will possibly be of different lengths.
+
+  """
+  if num_rounds == 0:
+    return mutations
+  mutations_to_combine = list(itertools.combinations(mutations, num_rounds + 1))
+
+  all_samples = []
+
+  for mutant in mutations_to_combine:
+    mutant = list(mutant)
+    prev_round = [()]
+    # iteratively add the mutations together
+    for i in range(num_rounds + 1):
+      next_round = []
+      new_mutation = mutant.pop()
+      for merged in prev_round:
```

Comment:
I don't follow why this step is necessary. mutations_to_combine already contains all possible combinations. Why do you need to do this exponentially big iterative thing inside the inner loop too?

Reply:
I see. It would be cleaner and easy to test and reason about if you employed a helper function for merging two sets. Then, you call this iteratively to merge ABC.